### PR TITLE
Adds deflate compression and fixes an http buffer bug

### PR DESCRIPTION
**Pull request by element-of-surprise**
This does two things:

Adds support for deflate compression from the client. ARN supports deflate and gzip.
Fixes what looks to me to be a possible race bug.  Was recycling a buffer before it was sent on the wire.

### CODE REVIEW

File: internal/conn/http/defalte_test.go
Comment (shanalily) [inline internal/conn/http/defalte_test.go:1]: nit: typo in file name

File: internal/conn/http/http.go [hunk 1]
```diff
@@ -42,11 +43,80 @@ var changeScope = map[string]bool{
...
+	},
+}
+
+// flateTransport is a custom RoundTripper that applies Deflate compression at the desired level.
+type flateTransport struct {
+	deflateLevel int
+	flatePool    chan *flate.Writer
+}
+
+func newFlatTransport() *flateTransport {
```
Comment (shanalily) [inline internal/conn/http/http.go:63]: I think this is also a typo? another nit, not necessary for merging

File: internal/conn/http/defalte_test.go [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+		w.Write([]byte("Request is not deflate-encoded"))
+	}
+}
+
+func TestDeflate(t *testing.T) {
+	t.Parallel()
+
+	data := make([]flateData, 0, 1000)
+
+	for i := 0; i < 1000; i++ {
```
Comment (shanalily) [inline internal/conn/http/defalte_test.go]: maybe this should test the empty request body case?

**Review by shanalily — APPROVED** [review]

File: internal/conn/http/defalte_test.go (hunk 2 above)
Comment (element-of-surprise) [reply internal/conn/http/defalte_test.go]: Added

**Pull request merged**